core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
     cachedSerializedStatuses.contains(shuffleId) || mapStatuses.contains(shuffleId)
   }
 
+  // Return the list of locations and blockSizes for each reducer.
+  // The map is keyed by reducerId and for each reducer the value contains the array
+  // of (location, size) of map outputs.
+  //
+  // This method is not thread-safe
+  def getStatusByReducer(
+      shuffleId: Int,
+      numReducers: Int)
```

Comment:
As discussed offline, take in a parameter here for maxNumberLocations, and then do the filtering for top 5 on line 309-ish when you create the status method (this avoids some of the scalability issues of creating a gigantic map)